modules/dnn/src/layers/depth_space_layer.cpp

```diff
@@ -0,0 +1,361 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+
+#include "../precomp.hpp"
+#include <opencv2/dnn/shape_utils.hpp>
+
+// OpenCL backend
+#ifdef HAVE_OPENCL
+#include "opencl_kernels_dnn.hpp"
+#endif
+
+// OpenVINO backend
+#ifdef HAVE_DNN_NGRAPH
+#include "../op_inf_engine.hpp"
+#include "../ie_ngraph.hpp"
+#endif
+
+// CUDA backend
+#ifdef HAVE_CUDA
+#include "../op_cuda.hpp"
+#include "../cuda4dnn/primitives/depth_space.hpp"
+#endif
+
+// CANN backend
+#ifdef HAVE_CANN
+#include "../op_cann.hpp"
+#endif
+
+// TIM-VX backend
+#ifdef HAVE_TIMVX
+#include "../op_timvx.hpp"
+#endif
+
+namespace cv { namespace dnn {
+
+class DepthSpaceLayerImpl CV_FINAL : public DepthSpaceLayer {
+public:
+
+    DepthSpaceLayerImpl(const LayerParams &params) {
+        setParamsFrom(params);
+
+        auto op_type = params.get<std::string>("op_type");
+        auto mode = params.get<std::string>("mode", "DCR");
+        CV_CheckFalse(op_type.empty(), "DepthSpaceLayer: op_type cannot be empty");
+        if (op_type == "DepthToSpace") {
+            if (mode == "DCR") {
+                op = OPERATION::DEPTH_TO_SPACE_DCR;
+                permutation = {0, 3, 4, 1, 5, 2};
+            } else if (mode == "CRD") {
+                op = OPERATION::DEPTH_TO_SPACE_CRD;
+                permutation = {0, 1, 4, 2, 5, 3};
+            } else {
+                CV_Error(Error::StsBadArg, cv::format("DepthSpaceLayer: unsupported mode %s\n", mode.c_str()));
+            }
+        } else if (op_type == "SpaceToDepth") {
+            op = OPERATION::SPACE_TO_DPETH;
+            permutation = {0, 3, 5, 1, 2, 4};
+        } else {
+            CV_Error(Error::StsBadArg, cv::format("DepthSpaceLayer: unsupported op_type %s\n", op_type.c_str()));
+        }
+
+        CV_CheckTrue(params.has("blocksize"), "DepthSpaceLayer: blocksize is required");
+        blocksize = params.get<int>("blocksize");
+    }
+
+    virtual bool supportBackend(int backendId) CV_OVERRIDE {
+        // TODO: support other backends
+#ifdef HAVE_DNN_NGRAPH
```

Comment:
ifdef can be avoided for OV and TimVX. dnn engine should manage check them earlier.

Reply:
Ok. Done.